source/material_model/visco_plastic.cc

```diff
@@ -0,0 +1,620 @@
+
+          // Composite viscosity
+          double viscosity_composite = (viscosity_diffusion * viscosity_dislocation)/(viscosity_diffusion + viscosity_dislocation);
+
+          // Select what form of viscosity to use (diffusion, dislocation or composite)
+          double viscosity_pre_yield = 0.0;
+          switch (viscous_type)
+            {
+              case diffusion:
+              {
+                viscosity_pre_yield = viscosity_diffusion;
+                break;
+              }
+              case dislocation:
+              {
+                viscosity_pre_yield = viscosity_dislocation;
+                break;
+              }
+              case composite:
+              {
+                viscosity_pre_yield = viscosity_composite;
+                break;
+              }
+              default:
+              {
+                AssertThrow( false, ExcNotImplemented() );
+                break;
+              }
+            }
+
+          // Calculate viscous stress
+          double viscous_stress = 2. * viscosity_pre_yield * edot_ii;
+
+          // Convert friction angle from degrees to radians
+          double phi = angles_internal_friction[j] * numbers::PI/180.0;
+
+          // Calculate drucker prager yield strength (i.e. yield stress)
+          double yield_strength = ( (dim==3)
+                                    ?
+                                    ( 6.0 * cohesions[j] * std::cos(phi) + 2.0 * std::max(pressure,0.0) * std::sin(phi) )
+                                    / ( std::sqrt(3.0) * (3.0 + std::sin(phi) ) )
+                                    :
+                                    cohesions[j] * std::cos(phi) + std::max(pressure,0.0) * std::sin(phi) );
+
+
+          // If the viscous stress is great than the yield strength, rescale the viscosity back to yield surface
+          double viscosity_yield = 0.0;
+          if ( viscous_stress >= yield_strength  )
+            {
+              viscosity_yield = yield_strength / (2.0 * edot_ii);
+            }
+          else
+            {
+              viscosity_yield = viscosity_pre_yield;
+            }
+
+          // Limit the viscosity with specified minimum and maximum bounds
+          composition_viscosities[j] = std::min(std::max(viscosity_yield, min_visc), max_visc);
+
+
+        }
+      return composition_viscosities;
+    }
+
+    template <int dim>
+    void
+    ViscoPlastic<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      for (unsigned int i=0; i < in.temperature.size(); ++i)
+        {
+          // const Point<dim> position = in.position[i];
+          const double temperature = in.temperature[i];
+          const double pressure= in.pressure[i];
+          const std::vector<double> composition = in.composition[i];
```

Comment:
A copy of the variables for temperature, pressure, and composition?  Yes, it could have be done without, by I do like the simplicity syntax when calling functions below.  I derived this portion of the code from @gassmoeller diffusion_dislocation material model and I think it would be helpful to keep such related portions of the code consistent for new users who want to develop there own material models.